src/padding.rs

```diff
@@ -1,12 +1,130 @@
-// Applies new padding to a BSON packet.
-// Returns the padded packet.
+use rand::Rng;
+use std::io::{Error, ErrorKind};
+
+/// Generates a vector of random bytes with the specified length, excluding a specific character.
+/// # Arguments
+/// * `rng` - A mutable reference to a random number generator.
+/// * `count` - The number of bytes to generate.
+/// * `exclusion` - The character to exclude from the generated bytes.
+/// # Returns
+/// A vector of random bytes with the defined length, excluding the specified character.
+fn generate_random_bytes<R: Rng>(rng: &mut R, count: u32, exclusion: char) -> Vec<u8> {
+    let mut bytes = Vec::with_capacity(count as usize);
+
+    // Use the random number generator to fill the defined buffer with random bytes.
+    let mut buf = [0u8; 256];
+    rng.try_fill_bytes(&mut buf)
+        .expect("Failed to fill buffer.");
+
+    // Iterate over the buffer to generate random bytes.
+    for &byte in &buf {
+        // Check if the desired number of bytes has been generated.
+        if bytes.len() == count as usize {
+            break;
+        }
+
+        // Exclude the specified character from the generated bytes
+        if byte != exclusion as u8 {
+            bytes.push(byte);
+        }
+    }
+
+    // Return the vector of random bytes.
+    bytes
+}
+
+/// Encapsulates the given BSON packet with a random number (between 2 and 32 inclusive) of random bytes before and after.
+/// # Arguments
+/// * `pkt` - A reference to a BSON packet.
+/// # Returns
+/// The encapsulated packet in Vec<u8> form.
+/// # Errors
+/// Returns an error if there is an error during the generation of random bytes or packet encapsulation
 fn encapsulate(pkt: &[u8]) -> Vec<u8> {
-    unimplemented!()
+    let mut rng = rand::thread_rng(); // use the random number generator to generate our random numbers and bytes
+    let r1 = rng.gen_range(2..=32); // length of random bytes before the BSON data
+    let r2 = rng.gen_range(2..=32); // length of random bytes after the BSON data
+
+    // vectors containing the random bytes for the front and back padding of the BSON data
+    let mut front_pad = generate_random_bytes(&mut rng, r1, '{');
+    let mut back_pad = generate_random_bytes(&mut rng, r2, '}');
+
+    front_pad.push('{' as u8);
+    back_pad.insert(0, '}' as u8);
+
+    // define a new vector to add the front pad, the packet, and finally the back pad
+    let mut encapsulated_pkt = Vec::new();
```

Comment:
This is something that should probably be done on aes too - but if we are going to use with_capacity somewhere, why not use it everywhere?